rp2040-hal/src/adc.rs

```diff
@@ -124,32 +187,49 @@ impl Channel<Adc> for TempSense {
     }
 }
 
-impl<WORD, PIN> OneShot<Adc, WORD, PIN> for Adc
+// Implementation for TempSense and type-checked pins
+impl<WORD, SRC> OneShot<Adc, WORD, SRC> for Adc
 where
     WORD: From<u16>,
-    PIN: Channel<Adc, ID = u8>,
+    SRC: Channel<Adc, ID = u8>,
 {
-    type Error = ();
-
-    fn read(&mut self, _pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
-        let chan = PIN::channel();
+    type Error = Infallible;
 
+    fn read(&mut self, _pin: &mut SRC) -> nb::Result<WORD, Self::Error> {
+        let chan = SRC::channel();
         if chan == TEMPERATURE_SENSOR_CHANNEL {
             self.device.cs.modify(|_, w| w.ts_en().set_bit())
         }
 
-        while !self.device.cs.read().ready().bit_is_set() {
-            cortex_m::asm::nop();
-        }
-
-        self.device
-            .cs
-            .modify(|_, w| unsafe { w.ainsel().bits(chan).start_once().set_bit() });
+        Ok(self.read(chan).into())
+    }
+}
 
-        while !self.device.cs.read().ready().bit_is_set() {
-            cortex_m::asm::nop();
-        }
+/// The pin was invalid for the requested operation
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct InvalidPinError;
 
-        Ok(self.device.result.read().result().bits().into())
+// Implementation for dyn-pins
+impl<WORD, F, M> OneShot<Adc, WORD, AdcPin<Pin<DynPinId, F, M>>> for Adc
```

Comment:
The ADC provides a u16, so does it make sense to be generic over any type WORD, where WORD can be created from a u16? e.g. if someone asked for a u32 from the read function, they'd get it, but it would only be scaled 0..65535 not 0..2^32. I think?

Reply:
🤷 That's how it was when I found it 🤷 I don't have opinions about it.
An ADC output may be 12 or 14bit and return the result on a u16 as 0..(2^12) and as far as I can remember it's not usually scaled up to a u16 so I guess one should not expect to get is scaled to a u32 even if that's the type they expect.